lib/pbench/server/api/resources/__init__.py

```diff
@@ -0,0 +1,694 @@
+
+
+class UnverifiedUser(SchemaError):
+    """
+    Unverified attempt to access other user data.
+    """
+
+    def __init__(self, username: str):
+        self.username = username
+        super().__init__(status=HTTPStatus.UNAUTHORIZED)
+
+    def __str__(self):
+        return f"{self.username} can not be verified"
+
+
+class InvalidRequestPayload(SchemaError):
+    """
+    A required client JSON input document is missing.
+    """
+
+    def __str__(self) -> str:
+        return "Invalid request payload"
+
+
+class UnsupportedAccessMode(SchemaError):
+    """
+    Unsupported values for user or access, or an unsupported combination of
+    both.
+    """
+
+    def __init__(self, user: str, access: str):
+        super().__init__()
+        self.user = user
+        self.access = access
+
+    def __str__(self) -> str:
+        return f"Unsupported mode {self.user}:{self.access}"
+
+
+class MissingParameters(SchemaError):
+    """
+    One or more required JSON keys are missing, or the values are unexpectedly
+    empty.
+    """
+
+    def __init__(self, keys: List[AnyStr]):
+        super().__init__()
+        self.keys = keys
+
+    def __str__(self):
+        return f"Missing required parameters: {','.join(self.keys)}"
+
+
+class ConversionError(SchemaError):
+    """
+    ConversionError Used to report an invalid parameter type
+    """
+
+    def __init__(self, value: Any, expected_type: str, actual_type: str):
+        """
+        Construct a ConversionError exception
+
+        Args:
+            value: The value we tried to convert
+            expected_type: The expected type
+            actual_type: The actual type
+        """
+        super().__init__()
+        self.value = value
+        self.expected_type = expected_type
+        self.actual_type = actual_type
+
+    def __str__(self):
+        return f"Value {self.value!r} ({self.actual_type}) cannot be parsed as a {self.expected_type}"
+
+
+class PostprocessError(Exception):
+    """
+    Used by subclasses to report an error during postprocessing of the
+    Elasticsearch response document.
+    """
+
+    def __init__(self, status: int, message: str, data: JSON = None):
+        self.status = status
+        self.message = message
+        self.data = data
+
+    def __str__(self) -> str:
+        return f"Postprocessing error returning {self.status}: '{str(self.message)} [{self.data}]'"
+
+
+def convert_date(value: str) -> datetime:
+    """
+    Convert a date/time string to a datetime.datetime object.
+
+    Args:
+        value: String representation of date/time
+        parameter: The Parameter definition (not used)
+
+    Raises:
+        ConversionError: input can't be validated or normalized
+
+    Returns:
+        datetime.datetime object
+    """
+    try:
+        return date_parser.parse(value)
+    except Exception:
+        raise ConversionError(value, "date/time string", type(value).__name__)
+
+
+def convert_username(value: str) -> str:
```

Comment:
Sigh; again, not directly relevant to the refactoring, but since this really belongs with the original ElasticBase refactoring, I guess I'll do it here.